File: bio-informatics/challenges/pattern-to-number.js
const { getNucleotideValue } = require('./nucleotide-utils');

const patternToNumber = (pattern, index = pattern.length - 1, total = 0) => {
    console.log({total})
    if (index < 0) { return total }

    const nucleotide = pattern[index];
    const value = getNucleotideValue(nucleotide);
    const binominalDecimal = pattern.length - ( index + 1);

    total += value * ( 4 ** binominalDecimal );

    return patternToNumber(pattern, index - 1, total);
};

module.exports = patternToNumber;

File: bio-informatics/challenges/python/utils/pattern_to_number.py
from utils import get_nucleotide_value

def pattern_to_number(pattern, index = None, total = 0):
  index = len(pattern) - 1 if index is None else index
  if index < 0:
    return total

  nucleotide = pattern[index]
    
  value = get_nucleotide_value(nucleotide)
  
  binominal_decimal = len(pattern) - ( index + 1)

  total += value * ( 4 ** binominal_decimal )

  return pattern_to_number(pattern, index - 1, total)
